[shard] make ShardedTensor a torch.Tensor subclass

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77592

This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

wanchaol committed:
[shard] make ShardedTensor a torch.Tensor subclass

This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36423976).

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36423976).

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -115,11 +114,47 @@ class ShardedTensor(object):
-        # Use __new__ for logging purposes.
+    _sharding_spec: shard_spec.ShardingSpec
+    _metadata: ShardedTensorMetadata
+
```
nit: Do we need slots?

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -790,183 +827,12 @@ def local_shards(self) -> List[Shard]:
-
-
-    def is_pinned(self) -> bool:
+    def is_pinned(self) -> bool:  # type: ignore[override]
```
Can't this be registered as a default tensor op too?

wanchaol replied:
So the issue with `is_pinned` is that it's actually getting dispatch to C++ and then to `torch_dispatch`, if we want to register it the way as default op, we will need to change `torch_dispatch` to handle this call in a special manner. I can think of how to do this in a follow up PR.

wanchaol committed:
Update on "[shard] make ShardedTensor a torch.Tensor subclass"


This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

Differential Revision: [D36423976](https://our.internmc.facebook.com/intern/diff/D36423976)

[ghstack-poisoned]

wanchaol committed:
Update on "[shard] make ShardedTensor a torch.Tensor subclass"


This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

Differential Revision: [D36423976](https://our.internmc.facebook.com/intern/diff/D36423976)

[ghstack-poisoned]

wanchaol committed:
Update on "[shard] make ShardedTensor a torch.Tensor subclass"


This is the reland of PR https://github.com/pytorch/pytorch/pull/74695, which was reverted due to some internal failures.

It also removes the ShardedTensorInterface change, we will delay that
change later if we found there's a need to do that.

Differential Revision: [D36423976](https://our.internmc.facebook.com/intern/diff/D36423976)

[ghstack-poisoned]
